### PR TITLE
A low scope error message height fix for #3091

### PR DESCRIPTION
**Pull request by xster**
Sorry for the dirt

### CODE REVIEW

File: packages/flutter/lib/src/rendering/error.dart [hunk 1]
```diff
@@ -9,6 +9,8 @@ import 'object.dart';
 
 const double _kMaxWidth = 100000.0;
 const double _kMaxHeight = 100000.0;
+/// This is a simplified estimate to offset the text from the top of the screen.
+const double _statusBarHeight = 24.0;
```
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: I assumed we want to minimize any dependencies here. Is it too much of a hack?

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: Part of the trouble is that the error widget can be anywhere in the tree.  If a widget throws an exception during the build, it turns into the error widget.  For that reason, we might want to move the text down 24.0 pixels because there might only be 24.0 pixels total in the widget.

What makes the issue tricky is just what you've identified: that we want to minimize dependencies because the system is likely to be in some crazy state when we're trying to show the error widget.

We've put off making a decision about what to do here for a while, but maybe it's time to make a better trade-off.  What we do currently does give a bad experience in some common cases.

I think it's fair to read ui.window.padding.top.  That's unlikely to be busted.  Let's move the text down if the height of the error object is at least 3x the top padding.  That won't do quite the right thing in the case where the object is clipped, but I think it's a better trade-off than our current behavior.

File: packages/flutter/lib/src/rendering/error.dart [hunk 2]
```diff
@@ -79,14 +81,15 @@ class RenderErrorBox extends RenderBox {
 
   /// The text style to use when painting [RenderErrorBox] objects.
   static ui.TextStyle textStyle = new ui.TextStyle(
-    color: const Color(0xFFFFFF00),
+    color: const Color(0xFFFFFF66),
     fontFamily: 'monospace',
-    fontSize: 7.0
+    fontSize: 10.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: there's a line height value that we should also set here

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: oh never mind i see you fixed it already

**Comment by Hixie** [comment]
You can't rely on the widget having a readable height iirc.

I think a better fix would be to print the error message three times, with a separator in between, something like:

```dart
final String line = '─' * 60;
final String message = '$error\n\n$line\n\n$error\n\n$line\n\n$error';
```

File: packages/flutter/lib/src/rendering/error.dart [hunk 3]
```diff
@@ -79,14 +81,15 @@ class RenderErrorBox extends RenderBox {
 
   /// The text style to use when painting [RenderErrorBox] objects.
   static ui.TextStyle textStyle = new ui.TextStyle(
-    color: const Color(0xFFFFFF00),
+    color: const Color(0xFFFFFF66),
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: why this change?

**Comment by Hixie** [comment]
Fixes https://github.com/flutter/flutter/issues/3091

**Comment by abarth** [comment]
That seems worth a try.

File: packages/flutter/lib/src/rendering/error.dart [hunk 4]
```diff
@@ -79,14 +81,15 @@ class RenderErrorBox extends RenderBox {
...
     fontFamily: 'monospace',
-    fontSize: 7.0
+    fontSize: 10.0,
+    fontWeight: FontWeight.bold
   );
 
   /// The paragraph style to use when painting [RenderErrorBox] objects.
   static ui.ParagraphStyle paragraphStyle = new ui.ParagraphStyle(
-    lineHeight: 0.25 // TODO(ianh): https://github.com/flutter/flutter/issues/2460 will affect this
+    lineHeight: 0.7 // TODO(ianh): https://github.com/flutter/flutter/issues/2460 will affect this
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: please remove this todo as well

File: packages/flutter/lib/src/rendering/error.dart (hunk 1 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: I like this solution. Done. 

I assume we can't make it ideal because the RenderBox doesn't know its absolute top position in the screen while painting?

File: packages/flutter/lib/src/rendering/error.dart (hunk 3 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Yellow on red wasn't the most contrasty for readability. It's still mostly the same color a bit brighter.

File: packages/flutter/lib/src/rendering/error.dart (hunk 4 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Done

File: packages/flutter/lib/src/rendering/error.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: Can we make the background darker and the foreground brighter? White on maroon, say? Or yellow on maroon?

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: Oh, I see, this is brighter. Never mind, I got confused.

File: packages/flutter/lib/src/rendering/error.dart [hunk 5]
```diff
@@ -104,6 +106,12 @@ class RenderErrorBox extends RenderBox {
           width = size.width;
         }
         _paragraph.layout(new ui.ParagraphConstraints(width: width));
+
+        // Heuristically offset the top of the text to prevent system status bar overlap if the
+        // widget is tall enough and likely to overlap.
+        if (size.height > 3 * systemTopPadding) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: nit: no braces

Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Done

File: packages/flutter/lib/src/rendering/error.dart [hunk 6]
```diff
@@ -50,6 +51,9 @@ class RenderErrorBox extends RenderBox {
 
   /// The message to attempt to display at paint time.
   final String message;
+  final double systemTopPadding = Zone.current['systemTopPadding'] == null
```
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: @abarth is this what you meant? Essentially dependency inject.

File: packages/flutter/test/rendering/error_test.dart [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+import 'rendering_tester.dart';
+
+class MockPaintingContext extends Mock implements PaintingContext {}
+class MockCanvas extends Mock implements Canvas {}
+
+/// Unit tests error.dart behaviors.
+void main() {
+  final String errorMessage = 'Some error message';
+  MockCanvas mockCanvas;
```
Comment (xster) [inline packages/flutter/test/rendering/error_test.dart]: I'll look into mock_canvas next.

File: packages/flutter/lib/src/rendering/error.dart (hunk 6 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: Yes, but I thinking we'd make this change in `dart:ui` so that the `ui.window` property was retrieved from the Zone.

```dart
final Window _window = new Window._();
Window get window => Zone.current[Window] ?? _window;
```

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: https://github.com/flutter/engine/blob/master/lib/ui/window.dart#L245 is where this code lives.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: Wait, what problem we solving here?

File: packages/flutter/lib/src/rendering/error.dart [hunk 8]
```diff
@@ -104,6 +109,11 @@ class RenderErrorBox extends RenderBox {
           width = size.width;
         }
         _paragraph.layout(new ui.ParagraphConstraints(width: width));
+
+        // Heuristically offset the top of the text to prevent system status bar overlap if the
+        // widget is tall enough and likely to overlap.
+        if (size.height > 3 * systemTopPadding)
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: If we do the three-line thing we don't have to do anyhting with the padding.

I'm not comfortable with us trying to read "size" here. This object is not in a sane state if it's being rendered.

File: packages/flutter/lib/src/rendering/error.dart (hunk 6 above)
Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: In testing, there's no way to control the values you get back from the `window` object.  For example, @xster wants to test how his render object behaves with different `window.padding.top` values.  Grabbing the window from the Zone lets you run your test in a Zone where you've substituted a different implementation of Window for `window`.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: Yeah but you also lose all your type checking.

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: We can keep the type checking if you like:

```dart
final Window _window = new Window._();
Window get window {
  final Object object = Zone.current[Window];
  return object is Window ? object : _window;
}
```

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: I'd just have the class have a protected getter that returns ui.window.whatever, and then in the test subclass that class and just override that getter.

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: Would you do that everywhere that reads from the `ui.window` static?  For example, the Zone approach would also let you test that WidgetApp correctly sets up the root MediaQueryData from the window:

https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/app.dart#L188

It would also let you test that the devicePixelRatio updates properly when the window notifies that its metrics have changed.

Another benefit of the Zone approach is that it will let you test code that someone else wrote that wasn't expecting to be tested in this way.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: I mean, fundamentally I would limit the number of places interacting with non-mockable global components like dart:ui so that this just doesn't really become a problem. Would you move all of dart:ui into a Zone?

I find the use of zones for this kind of thing to be very aesthetically unsatisfactory from an API perspective. Also, in addition to the loss of strong typing (by which I specifically mean that the analyzer can't verify that code reading out of zones in this way is correct), you lose dartdocs and autocompletion.

> Another benefit of the Zone approach is that it will let you test code that someone else wrote that wasn't expecting to be tested in this way.

Not really, because someone who isn't expecting to be testing in this way wouldn't use zones (unless you mean specifically for things we are forcing them to use zones for, but then it seems like we're the ones who are expecting things to be tested this way, if you see what I mean).

The flutter_tools code uses zones a bit for this. I don't think in practice that it's been a win over more straight-forward injection.

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: > Would you move all of dart:ui into a Zone?

We only have one static, so there isn't much else to move.  :)

> Also, in addition to the loss of strong typing (by which I specifically mean that the analyzer can't verify that code reading out of zones in this way is correct), you lose dartdocs and autocompletion.

I don't quite understand this part.  I might need to ask you to explain in person.

> unless you mean specifically for things we are forcing them to use zones

Yes, I meant specifically for the `window` static.

Ok, I'm happy with the `@protected` member function approach you suggest.  If we find ourselves doing that in many places, we can re-evaluate whether we want a more centralized solution.

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: I see, you mean on the setting side.  Thank makes sense.  Thanks.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: I mean anywhere that you would currently type `ui.window` and the analyzer would know it was a Window, the autocomplete would know it had a `padding` member, the VM would verify the type, etc. Setting side and getting side. You lose all of that with zones.

We have more than one static in dart:ui. For example, you could use zones to mock all the methods, all the getters and setters for callbacks, all the functions, etc. Suppose you wanted to test that a particular hashCode function called hashValues, for example. You could by putting hashValues into a Zone.current map and requiring that it be accessed that way. Similarly if we put the constructors for Path and company into Zone.current, we could provide mocks for those classes.

I think if we wanted to go down a route like that we should prefer to ask for explicit support from Dart for mocking a library. It would presumably be reasonably straight-forward for the VM team to provide a way to substitute another library in any time we `import 'dart:ui'`. This would let us inject an entire mock library, which could re-export from the "real" dart:ui the parts that it doesn't want to mock, providing only the specific parts it wants to override.

Anyway I don't want to beat a dead horse so I'll shut up now. :-)

Comment (abarth) [inline packages/flutter/lib/src/rendering/error.dart]: > You lose all of that with zones.

On the getter side, you can keep the `ui.window` getter with the appropriate type (as in the code snippet above) keep all that good stuff.  The setting side is tricker.

> Anyway I don't want to beat a dead horse.

Hopefully we haven't scared @xster :)

Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Processing the info a bit :)
My 2 cents:
- In the very long term, I mostly agree with Ian's aesthetic views. Avoiding global statics will likely save us a bunch of trouble down the road. Along the axis of strict disciplined API <-> Python-like magical experience, using something like https://github.com/angular/di.dart or letting RenderObjects (maybe?) retrieve the Window object from the this. instance rather than from a global is probably more predictable. Everyone reading the getter and the setter knows it's overridable. Zone getter overrides is probably in the middle since it seems like a flat dictionary of overrides. Import overrides is probably more magical. Works in this case but if promoted as a generic pattern, developers could end up with a number of surprises. 

- For this very particular case, we may be over engineering this. If we add smarts to window padding, we might not want to run it for every 16ms, especially considering the dependency reduction we might want to have on this file. If it's not smart, we might as well shed the dependency entirely and hard-code an offset.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: For this case we don't need the padding at all. We should just duplicate the message three times as per the comment below and thus remove all need to worry about the padding.

**Comment by xster** [comment]
Ok, reverted various fanciness. Prints the error message multiple times. PTAL.

File: packages/flutter/lib/src/rendering/error.dart [hunk 9]
```diff
@@ -9,6 +9,7 @@ import 'object.dart';
 
 const double _kMaxWidth = 100000.0;
 const double _kMaxHeight = 100000.0;
+const String line = '\n──────────────────────\n\n';
```
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: It's this long without trying to read from size so it doesn't line wrap on small phones like iPhone SE

Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: Add a comment to that effect. Also, have you checked what it looks like when the exception happens inside a 100x100 widget? Or a 10x200 widget, or a 200x10 widget? :-)

File: packages/flutter/lib/src/rendering/error.dart [hunk 10]
```diff
@@ -42,7 +43,9 @@ class RenderErrorBox extends RenderBox {
         // see the paragraph.dart file and the RenderParagraph class.
         ui.ParagraphBuilder builder = new ui.ParagraphBuilder(paragraphStyle);
         builder.pushStyle(textStyle);
-        builder.addText(message);
+        builder.addText(
+          '$message$line$message$line$message$line$message$line$message$line$message$line'
+          '$message$line$message$line$message$line$message$line$message$line$message');
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/error.dart]: nit: newline before `)`

**Comment by Hixie** [comment]
This looks great.

What was your conclusion regarding testing this?

LGTM, please include a test if you can find a way to test it. Otherwise don't worry.

File: packages/flutter/lib/src/rendering/error.dart (hunk 10 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Done

**Comment by xster** [comment]
Brought back error_test with mock_canvas. 
Added a paragraph matcher. Though the paragraph assertion isn't as interesting given the current API if I understood correctly. Probably need to expose another API from the engine for a text getter or some other form of == override I imagine.

File: packages/flutter/lib/src/rendering/error.dart (hunk 9 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Looks still legible in all cases though printing multiple times overflows the parent container

File: packages/flutter/lib/src/rendering/error.dart (hunk 6 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Done

File: packages/flutter/test/rendering/error_test.dart (hunk 7 above)
Comment (xster) [inline packages/flutter/test/rendering/error_test.dart]: Done

File: packages/flutter/lib/src/rendering/error.dart (hunk 8 above)
Comment (xster) [inline packages/flutter/lib/src/rendering/error.dart]: Done
